### PR TITLE
Builds rust files for the allocation module

### PR DESCRIPTION
**Pull request by zmanian**
Uses a similar strategy to what we in the gravity repo to build rust bindings for the proto files in sommelier repo.

### CODE REVIEW

File: prost_build/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+use walkdir::WalkDir;
+
+/// Protos belonging to these Protobuf packages will be excluded
+/// (i.e. because they are sourced from `tendermint-proto` or `cosmos-sdk-proto`)
+const EXCLUDED_PROTO_PACKAGES: &[&str] = &["gogoproto", "google", "tendermint", "cosmos.base"];
+/// Regex for locating instances of `cosmos-sdk-proto` in prost/tonic build output
+const COSMOS_SDK_PROTO_REGEX: &str = "(super::)+cosmos";
+
+/// A temporary directory for proto building
+const TMP_PATH: &str = "/tmp/sommelier/";
```
Comment (levicook) [inline prost_build/src/main.rs:24]: are there any standard lib facilities for creating a safe temp dir? (generally random-ish name under $TMP)

Comment (levicook) [reply prost_build/src/main.rs:24]: (asking more out of curiosity than need)

File: prost_build/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+// working directory.
+
+fn main() {
+    let out_path = Path::new(&OUT_PATH);
+    let tmp_path = Path::new(&TMP_PATH);
+    compile_protos(out_path, tmp_path);
+}
+
+fn compile_protos(out_dir: &Path, tmp_dir: &Path) {
+    println!(
```
Comment (levicook) [inline prost_build/src/main.rs:38]: nit: better for informational stuff flow out of stderr / eprintln to help people that pipe stuff (me)

File: prost_build/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    // `prost-build` output references types from `tendermint-proto` crate
+    // relative paths, which we need to munge into `tendermint_proto` in
+    // order to leverage types from the upstream crate.
+    let contents = Regex::new(COSMOS_SDK_PROTO_REGEX)
+        .unwrap()
+        .replace_all(&contents, "cosmos_sdk_proto::cosmos");
+
+    fs::write(dest, contents.as_bytes())
+}
```
Comment (levicook) [inline prost_build/src/main.rs:155]: this is a port of the first Rust code I ever read and am glad to see it makes 1000x more sense now

**Review by levicook — APPROVED** [review]
one question, one nit, one non-sequitur -- nothing to worry over 🚢

File: prost_build/src/main.rs (hunk 1 above)
Comment (zmanian) [reply prost_build/src/main.rs:24]: Yeah this exists

https://doc.rust-lang.org/std/env/fn.temp_dir.html

**Pull request merged**